qutebrowser/browser/history.py

```diff
@@ -71,162 +71,78 @@ def url_str(self):
-                             Used to sync with the completion.
-                             arg: The new Entry.
-        item_added: Emitted after a new Entry is added.
-                    Used to tell the savemanager that the history is dirty.
-                    arg: The new Entry.
-        cleared: Emitted after the history is cleared.
-    """
-
-    add_completion_item = pyqtSignal(Entry)
-    item_added = pyqtSignal(Entry)
-    cleared = pyqtSignal()
-    async_read_done = pyqtSignal()
-
-    def __init__(self, hist_dir, hist_name, parent=None):
-        super().__init__(parent)
-        self._initial_read_started = False
-        self._initial_read_done = False
-        self._lineparser = lineparser.AppendLineParser(hist_dir, hist_name,
-                                                       parent=self)
-        self.history_dict = collections.OrderedDict()
-        self._temp_history = collections.OrderedDict()
-        self._new_history = []
-        self._saved_count = 0
-        objreg.get('save-manager').add_saveable(
-            'history', self.save, self.item_added)
+    def __init__(self, parent=None):
+        super().__init__("History", ['url', 'title', 'atime', 'redirect'],
+                         parent=parent)
+        self.completion = CompletionHistory(parent=self)
+        self.create_index('HistoryIndex', 'url')
+        self._contains_query = self.contains_query('url')
+        self._between_query = sql.Query('SELECT * FROM History '
+                                        'where not redirect '
+                                        'and not url like "qute://%" '
+                                        'and atime > ? '
+                                        'and atime <= ? '
+                                        'ORDER BY atime desc')
+
+        self._before_query = sql.Query('SELECT * FROM History '
+                                       'where not redirect '
+                                       'and not url like "qute://%" '
+                                       'and atime <= ? '
+                                       'ORDER BY atime desc '
+                                       'limit ? offset ?')
 
     def __repr__(self):
         return utils.get_repr(self, length=len(self))
 
-    def __iter__(self):
-        return iter(self.history_dict.values())
-
-    def __len__(self):
-        return len(self.history_dict)
+    def __contains__(self, url):
+        return self._contains_query.run([url]).value()
 
-    def async_read(self):
-        """Read the initial history."""
-        if self._initial_read_started:
-            log.init.debug("Ignoring async_read() because reading is started.")
-            return
-        self._initial_read_started = True
+    def _add_entry(self, entry):
+        """Add an entry to the in-memory database."""
+        self.insert([entry.url_str(), entry.title, int(entry.atime),
+                     entry.redirect])
+        if not entry.redirect:
```

Comment:
We dropped the check for `qute://` URLs along the way, we might want to reintroduce it?

(also, @lahwaacz mentioned: "I was actually thinking that only `qute://history` pages should be ignored, showing pages like `qute://help` in the history (and even completion later?) would be helpful. But since the current behaviour is to ignore all `qute://` pages in the history, it should be changed in a different PR." - just so we don't forget)